[macOS] Implement merged UI and platform thread

knopp opened the pull request:
Original issue: https://github.com/flutter/flutter/issues/150525

This PR lets the macOS embedder run both with and without UI and platform thread merged.

Thread merging is controlled through `FLTEnableMergedPlatformUIThread` `info.plist` option similar to iOS embedder, though the default value is currently `false`.

Changes in the resize / vsync synchronization:

- Added `FlutterRunLoop` class to schedule Flutter tasks on main thread in a way where it is possible to only process these (Flutter posted) tasks while waiting for correct frame size during resizing. This significantly simplifies the resize synchronization and makes the same code work both with separate UI thread and with UI and platform thread merged.
- `FlutterThreadSynchronizer` has been renamed to `FlutterResizeSynchronizer` vastly simplified, mutex and conditions are removed and the blocking is now done by only processing Flutter messages while waiting for resizing. It is now per view (instead of storing a viewId->Size map internally) and owned by the view itself, instead of engine.
- This approach to resize synchronization will work for Windows and Linux as well. This will allow us to conceptually consolidate the way we do threading and resize synchronization on all three desktop embedders.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

knopp committed:
Allow running macOS embedder with UI and platform thread merged

munrocket commented:
Awesome! Making own graphic API with FFI on macOS is impossible without this. We need to land it ASAP.

knopp commented:
Putting this as WIP. With this PR with unmerged thread, CVDisplayLink is started on UI thread and stopped on platform thread. I initially assumed that would work, but it doesn't. During resizing the CVDisplayLink creates two threads and trying to stop it deadlocks. 

CVDisplayLink API documentation doesn't make any threading guarantees, I assumed it was thread safe but it doesn't seem to be the case.

knopp committed:
Simplify vsync waiter threading

knopp commented:
I've simplified display link / vsync waiter threading model. Now everything is expected to happen on main thread. This requires one extra hop from UI to platform thread with unmerged threads, and zero extra hops when running with merged threads (which will become default).

The mental overhead of trying to keep everything thread safe was probably not really worth it with unmerged threads and is definitely not worth it with merged threads. Also while not documented anywhere, `CVDisplayLink` seems to break when start and stop are called from different threads (despite having an internal mutex). This is no longer an issue since all start / stop calls are serialized and performed on main thread.

johnmccutchan commented:
@knopp This is so great!

loic-sharma commented:
cc @dkwingsmt as you've also worked on the macOS embedder

jonahwilliams commented:
@knopp sorry we lost track of this, should I start reviewing?

knopp committed:
Include <atomic>

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterDisplayLink.mm:
```diff
@@ -61,68 +45,40 @@ void OnDisplayLink(CVDisplayLinkRef display_link,
 
 void DisplayLinkManager::UnregisterDisplayLink(_FlutterDisplayLink* display_link) {
-  std::unique_lock<std::mutex> lock(mutex_);
+  FML_DCHECK([NSThread isMainThread]);
```
nit (here and below): for new code we should use `NSThread.isMainThread`.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterDisplayLinkTest.mm:
```diff
@@ -33,38 +34,45 @@ - (void)onDisplayLink:(CFTimeInterval)timestamp targetTimestamp:(CFTimeInterval)
   [window setContentView:view];
 
-  auto event = std::make_shared<fml::AutoResetWaitableEvent>();
+  __block bool signalled = false;
```
nit, here and below: Unless we're passing to C++ calls that take a `bool`, prefer `BOOL` / `YES` (below).

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -676,10 +677,24 @@ - (BOOL)runWithEntrypoint:(NSString*)entrypoint {
           },
   };
+
+  bool mergedPlatformUIThread = false;
```
nit: go for `BOOL`/`NO` here.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:
```diff
@@ -1540,29 +1547,21 @@ - (BOOL)unregisterTextureWithID:(int64_t)textureID {
+          }
+        }
+      }
+        afterDelay:(targetTime - (double)engine_time) / 1000000000.0];
```
Use [`NSEC_PER_SEC`](https://developer.apple.com/documentation/dispatch/nsec_per_sec?language=objc) from Foundation.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:
```diff
@@ -526,9 +525,14 @@ @implementation MockableFlutterEngine
 
-  [engine.testThreadSynchronizer blockUntilFrameAvailable];
-
+  CFAbsoluteTime start = CFAbsoluteTimeGetCurrent();
```
```suggestion
  // Wait up to 1 second for Flutter to emit a frame.
  CFAbsoluteTime start = CFAbsoluteTimeGetCurrent();
```

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:
```diff
@@ -980,7 +985,15 @@ @implementation MockableFlutterEngine
   EXPECT_TRUE([engine runWithEntrypoint:@"drawIntoAllViews"]);
-  [engine.testThreadSynchronizer blockUntilFrameAvailable];
+
+  CFTimeInterval start = CACurrentMediaTime();
```
```suggestion
  // Wait up to 1 second for Flutter to emit a frame.
  CFTimeInterval start = CACurrentMediaTime();
```

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterResizeSynchronizer.h:
```diff
@@ -0,0 +1,40 @@
+#import "flutter/shell/platform/darwin/macos/framework/Headers/FlutterViewController.h"
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.h"
+
+@interface FlutterResizeSynchronizer : NSObject
```
Nit: Add a short summary indicating this is for coordinating window resizes.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterResizeSynchronizer.mm:
```diff
@@ -0,0 +1,61 @@
+@implementation FlutterResizeSynchronizer {
+  std::atomic_bool _inResize;
+  bool _shuttingDown;
+  bool _didReceiveFrame;
```
`BOOL` for these two since you set both to `YES` below.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterResizeSynchronizerTest.mm:
```diff
@@ -0,0 +1,119 @@
+  __block BOOL commit2 = NO;
+
+  fml::AutoResetWaitableEvent latch_;
+  fml::AutoResetWaitableEvent& latch = latch_;
```
I'm assuming the need for the reference has something to do with usage in a block, and not being able to mark it `__block` for some reason? If so, please add a comment (similar for the one below) -- that's a new one to me!

knopp replied:
Capturing c++ objects requires copy constructor (which AutoResetWaitableEvent) does not have. What __block does is move this copy to the heap. If still needs copy constructor. Capturing reference on the other hand seems just like capturing a pointer so it works with AutoResetWaitableEvent.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.mm:
```diff
@@ -0,0 +1,130 @@
+#import "flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.h"
+#include "fml/logging.h"
```
nit: move this below `<vector>` to make the style guide gods happy. 

Groups are:
1. Associated header
2. std C headers
3. std C++ headers
4. other headers

The google style guide splits 4 into two groups, but that's just crazy talk.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.mm:
```diff
@@ -0,0 +1,130 @@
+  CFAbsoluteTime target_time;
+};
+
+CFStringRef kFlutterRunLoopMode = CFSTR("FlutterRunLoopMode");
```
```suggestion
const CFStringRef kFlutterRunLoopMode = CFSTR("FlutterRunLoopMode");
```

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.mm:
```diff
@@ -0,0 +1,130 @@
+
+}  // namespace
+
+@implementation FlutterRunLoop (Private)
+
+CFRunLoopRef _runLoop;
+CFRunLoopSourceRef _source;
+CFRunLoopTimerRef _timer;
+std::vector<Task> _tasks;
+
```
Shouldn't these be in an ivar block? I'm actually kind of surprised that this compiles; I wonder if they're being treated as globals?

```suggestion
@implementation FlutterRunLoop (Private) {
  CFRunLoopRef _runLoop;
  CFRunLoopSourceRef _source;
  CFRunLoopTimerRef _timer;
  std::vector<Task> _tasks;
}
```

knopp replied:
Ouch.

cbracken commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.mm:
```diff
@@ -0,0 +1,130 @@
+
+- (void)performBlock:(void (^)(void))block afterDelay:(NSTimeInterval)delay {
+  @synchronized(self) {
+    _tasks.push_back({block, CFAbsoluteTimeGetCurrent() + delay});
```
nit: I _think_ in this case you can get away with `emplace_back`:
```suggestion
    _tasks.emplace_back(block, CFAbsoluteTimeGetCurrent() + delay);
```

knopp replied:
Seems to compile...

cbracken reviewed (approved):
LGTM modulo a few nits!

![LGTM stamp from a Japanese personal seal](https://github.com/user-attachments/assets/7e9438cf-17df-40c1-8594-f9720054d7d2)

knopp committed:
Nits

knopp committed:
Comment

knopp committed:
Nit

knopp committed:
Fix double release because of shallow runner description copy

The descriptions share single bridge retained engine while both of them would release it.

knopp committed:
Add missing shutDown

munrocket commented:
Wow, I thought Flutter is dead ☠️
Now we can play with custom graphics on MacOS.
